### DISCUSSION TITLE
Running the application by command

### DISCUSSION THREAD

**Discussion by BurningFiber** [K-Society/KSociety.SharpCubeProgrammer#144] [Q&A]
Hey there,
first of all thanks for that awesome library. Programming via UART works really great.

After downloading the hex-file I always need to restart the controller manually. I hoped that I may start the programm automatically after flashing like this:

```
var downloadFile = cubeProgrammerApi.DownloadFile(firmwarePath, "0x08000000", 0U, 1U);
var execute = cubeProgrammerApi.Execute("0x08000000");
```

But nothing will happen after the execute command. Has that function been already tested? Any idea what I am missing?

This is the log when executing the file:

```
RUNNING Program ... 
  Address:      : 0x08000000
Sending Start command and its XOR:
byte 0x21 sent successfully to target
byte 0xDE sent successfully to target
Wait ends after 1 loop, dataready = 1, delay = 1
Received response from target: 0x79
Sending start address and its checksum:
data sent successfully to target: 0x0800000008
Wait ends after 1 loop, dataready = 1, delay = 1
Received response from target: 0x79
Start operation achieved successfully
```

**Comment by maniglia** [comment]
Hello @BurningFiber 
Thanks for your feedback, I've never tested programming via UART yet.
However, both functions (_DownloadFile_ and _Execute_) are the same regardless of the type of connection used.
What happens if you try to program using CubeProgrammer software?
Do you set the _Option Bytes_ directly from the firmware?
If so, then it is correct that the microcontroller does not start, but you can try to set the Option Bytes, using this library, before programming the micro.

**Comment by BurningFiber** [comment]
So far I just programmed the microcontrollers using the STM32CubeProgrammer Software (Windows). There is an option to run the programm after programming and that is working fine if I check the checkbox:

![Screenshot 2024-06-21 095929](https://github.com/K-Society/KSociety.SharpCubeProgrammer/assets/88766777/5834a3e2-c5e6-4e21-ac33-f48c62b92ec9)

I hoped that this would be possible also using your library. I did not set any Option Bytes in STM-Software or in my C#-Project. Sadly I don't know if there are any Option Bytes set in the firmware since I just got the firmware from an external supplier that is not available in the moment.

Is there any Option Byte that I need to send via your library?

**Comment by maniglia** [comment]
In theory, everything that can be done with the CubeProgrammer software can also be done with CubeProgrammer_API and consequently with this library.
I think the given option (Run after programming) is the equivalent of the _Execute_ command, there is no other command to start the microcontroller.
If the _Execute_ command returns _Success_ but nothing happens after that, it makes me think that there is some problem that doesn't depend on the library, maybe the firmware starts but stops immediately afterwards.
This should also happen with the CubeProgrammer software.
What you can do is compare the log messages of the CubeProgrammer software and your C# application, they should be very similar.
To say that the microcontroller doesn't start I imagine that after programming you try to interact with it by sending it some commands or maybe wait for some LEDs to turn on. 
The _Option Bytes_ are defined and set by the person who wrote the firmware, such as _Readout Protection_, if the firmware checks and sets the _Readout Protection_ at startup then a hardware reboot is required, even with the CubeProgrammer software.
Last question, is the address "0x08000000" correct?
Unfortunately I have no experience with programming via UART.

**Reply by BurningFiber** [reply]
Please excuse my late reply. The last weeks I moved with my office to another location.

> To say that the microcontroller doesn't start I imagine that after programming you try to interact with it by sending it some commands or maybe wait for some LEDs to turn on.

Yes I have a display and some interfaces connected to the STM that directly show if the programm is working.

> Last question, is the address "0x08000000" correct?

Yes that is right.

I just resumed my troubleshooting and found out that "Run after programming" also does **not** work in the Windows STM32CubeProgrammer with UART Connection. It only works with the ST-LINK, which should be able to reset the STM32 via the NRTS-Pin to leave the bootloader. So I think it will not be an issue with your library but a problem that it's generally not possible to trigger the reset in UART-Mode. I am not that firm in microcontrollers but I think our firmware needs the reset to start the programm. Maybe somebody has further information but I think I will continue without automatic start of the program. 

Thank you for your help!
